src/meta/processors/job/BalanceJobExecutor.cpp

```diff
@@ -665,174 +439,327 @@ void BalanceJobExecutor::calDiff(const HostParts& hostParts,
     }
-
-    LOG(INFO) << "source " << source << " h.first " << h.first;
-    if (std::find(it->second.begin(), it->second.end(), partId) == it->second.end()) {
-      return h.first;
+    if (status == meta::cpp2::JobStatus::FINISHED) {
+      nebula::cpp2::ErrorCode ret = updateMeta();
+      if (ret != nebula::cpp2::ErrorCode::SUCCEEDED) {
+        status = meta::cpp2::JobStatus::FAILED;
+      }
     }
+    onFinished_(status);
+  });
+  plan_->invoke();
+  return Status::OK();
+}
+
+nebula::cpp2::ErrorCode ZoneBalanceJobExecutor::updateMeta() {
+  std::string spaceKey = MetaKeyUtils::spaceKey(spaceInfo_.spaceId_);
+  std::string spaceVal;
+  kvstore_->get(kDefaultSpaceId, kDefaultPartId, spaceKey, &spaceVal);
+  meta::cpp2::SpaceDesc properties = MetaKeyUtils::parseSpace(spaceVal);
+  std::vector<std::string> zones;
+  for (std::string& zn : lostZones_) {
+    spaceInfo_.zones_.erase(zn);
+  }
+  for (auto& p : spaceInfo_.zones_) {
+    zones.push_back(p.first);
   }
-  return nebula::cpp2::ErrorCode::E_NO_HOSTS;
+  properties.set_zone_names(std::move(zones));
+  std::vector<kvstore::KV> data;
+  data.emplace_back(MetaKeyUtils::spaceKey(spaceInfo_.spaceId_),
+                    MetaKeyUtils::spaceVal(properties));
+  folly::Baton<true, std::atomic> baton;
+  auto ret = nebula::cpp2::ErrorCode::SUCCEEDED;
+  kvstore_->asyncMultiPut(kDefaultSpaceId,
+                          kDefaultPartId,
+                          std::move(data),
+                          [&baton, &ret](nebula::cpp2::ErrorCode code) {
+                            if (nebula::cpp2::ErrorCode::SUCCEEDED != code) {
+                              ret = code;
+                              LOG(ERROR) << "Can't write the kvstore, ret = "
+                                         << static_cast<int32_t>(code);
+                            }
+                            baton.post();
+                          });
+  baton.wait();
+  return ret;
 }
 
-bool DataBalanceJobExecutor::checkZoneLegal(const HostAddr& source, const HostAddr& target) {
-  VLOG(3) << "Check " << source << " : " << target;
-  auto sourceIter = std::find_if(zoneParts_.begin(), zoneParts_.end(), [&source](const auto& pair) {
-    return source == pair.first;
-  });
+/* first, move the lostZones' parts to the active zones
+ * second, make balance for the active zones */
+Status ZoneBalanceJobExecutor::buildBalancePlan() {
+  for (std::string& zn : lostZones_) {
+    if (!spaceInfo_.zones_.count(zn)) {
+      return Status::Error("space %s does not have zone %s", spaceInfo_.name_.c_str(), zn.c_str());
+    }
+  }
 
-  if (sourceIter == zoneParts_.end()) {
-    LOG(INFO) << "Source " << source << " not found";
-    return false;
+  std::map<std::string, Zone*> activeZones;
+  std::map<std::string, Zone*> lostZones;
+  for (auto& zoneMapEntry : spaceInfo_.zones_) {
+    activeZones.emplace(zoneMapEntry.first, &zoneMapEntry.second);
   }
+  for (std::string& zn : lostZones_) {
```

Comment:
now activeZones is not ready yet. I check after activeZones is ready

Reply:
You have a local variable named `activeZones` here, from the space val and parameter could get it just like what you did. This variable is enough?

Reply:
> You have a local variable named `activeZones` here, from the space val and parameter could get it just like what you did. This variable is enough?

what's the problem?

Reply:
For example, if a space has 4 related zones, but use ask to remove 2 zones in the request, then we should report error instead of continuing to build the plan